cython, numpy: build for mixed Python versions

iMichka opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is inspired by what we did for `Pillow`. The idea is to allow mixed Python dependencies. We can later on extend the idea to `scipy`, I just wanted to start small.
This change should ease the Python 3.10 migration, as both formulae are quite low in the dependency tree, and my hope is to be able to decouple things a little bit, to be able to handle bits of the dependency tree separately.

I still have one bug with `numpy` and Python 3.10, the build currently fails; I need to check that.
We might also need to add some `python@xy` dependencies to formulae that use `cython` or `numpy` as these will be missing now. I shifted the responsibility to choose which `python@xy` formula to install to the upper formulae in the dependency tree.

This is a draft and needs further testing / thinking / tweaking. I would like some feedback on the strategy too.

iMichka commented:
This builds now and should be good; I just need to fix the audit issue.

How can I disable the following audit?
>Formula/cython.rb:23:14: C: [Correctable] References to python@3.8 should match the specified python dependency (python@3.10)
    depends_on "python@3.8" => [:build, :test]

@Homebrew/core for general thoughts about this PR?

carlocab commented on Formula/cython.rb:
```diff
@@ -19,21 +20,32 @@ class Cython < Formula
-  depends_on "python@3.9"
+  depends_on "python@3.8" => [:build, :test]
+  depends_on "python@3.9" => [:build, :test]
+  depends_on "python@3.10"
```
The idea of this PR is fine with me, though I don't think there is a way around the audit failure while `python@3.10` is a runtime dep.

iMichka replied:
Yeah initially I thought I could make it a build/test dep, bit it is still needed for the cython binary.

I can add a new allowlist in brew I guess?

MikeMcQuaid replied:
> I can add a new allowlist in brew I guess?

Yes, please!

carlocab replied:
Another alternative to this, in the case of `cython`, since it ships executables, is to provide separate Cython formulae which each depend on a different version of Python.

I'm of two minds about this: this would lead to multiple formulae which build from the same source. On the other hand, it might provide a better user experience (e.g. we could make at least one of these formulae not-keg-only since the keg-only status of `cython` is really from the time we didn't want to package anything you could `pip install`).

carlocab replied:
To be clear on this, though: I am definitely against multiple `numpy` formulae, since it does not ship executables, and the builds that result from different Python versions do not conflict.

iMichka commented:
> Another alternative to this, in the case of cython, since it ships executables, is to provide separate Cython formulae which each depend on a different version of Python.

I think I am going to implement this. I will update the PR accordingly.

On thing that bothers me though, is that the formulae will be called `cython@3.8`,  `cython@3.9`,  `cython@3.10`, where the version number corresponds to the Python version, and not the `cython` version. But I think we can make that exceptions as it is pretty clear what is going on with these version numbers.

carlocab commented:
I'd rather not use `@` here. `@` should refer to the formula's version. It may be clear to us that it won't for the Cython formulae, but I don't think it will for many users.

iMichka commented:
So rather `cython38`, `cython39`, `cython310` ?

MikeMcQuaid commented:
> This change should ease the Python 3.10 migration, as both formulae are quite low in the dependency tree, and my hope is to be able to decouple things a little bit, to be able to handle bits of the dependency tree separately.

If this is the primary motivation: what versions do we actually need for that?

iMichka commented:
So, my main focus is `numpy` right now, as it is quite a central dependency.

If you execute `python3.9` and the try to `import numpy`, this will work as this is what we ship.
But if you run  `python3.10` or  `python3.8`, it won't work as we don't install numpy to these Python's site-package folders.

It should at least work for Python 3.9 and 3.10, because this is what we need for the migration.
But I plan to add support for 3.8 and 3.7 too, as these are the Python versions we ship and that are still supported. Python 3.11 is coming soon so this should even add a fifth one.

This means we need at least 2 `cython` formulae, one built upon Python 3.9 and one for Python 3.10. But as said before, we would have 4 `cython` formulae, one for each Python version, to be consistent and provide a working `numpy` installation for each of our Python versions.

Note that these `cython` formulae are still only for internal brew usage. Maintaining them is not a big deal and should not have a huge cost.

MikeMcQuaid commented:
> But if you run `python3.10` or `python3.8`, it won't work as we don't install numpy to these Python's site-package folders.
> 
> It should at least work for Python 3.9 and 3.10, because this is what we need for the migration. But I plan to add support for 3.8 and 3.7 too, as these are the Python versions we ship and that are still supported. Python 3.11 is coming soon so this should even add a fifth one.

I think we should ship only what we need for the migration, no more than that and remove these afterwards. I don't want to get into the habit of us having parallel dependency trees for versioned formulae. In general: I don't really like this approach at all but I'm open to being more pragmatic if it makes our migration easier.

> This means we need at least 2 `cython` formulae, one built upon Python 3.9 and one for Python 3.10. But as said before, we would have 4 `cython` formulae, one for each Python version, to be consistent and provide a working `numpy` installation for each of our Python versions.

Yeh, I don't think we should do <3.9, as above. I also think, if they aren't already, the versioned ones should be keg-only.

iMichka commented:
I reduced the scope to the bare minimun, as asked: Python/Cython 3.10 and 3.9 only.

I did not touch the original `cython` formula. Once I have migrated all formulae to `cython-39` and `cython-310`, we can drop the unversioned `cython` formulae completely. Both new formulae are for internal usage only, if someone needs `cython`, they should use `pip`, as stated in the keg only reason.

Note that for `numpy`, there is no Python runtime dependency anymore. I consider `numpy` as an internal library, as it is only there to help build some formulae that need `numpy` and reduce vendoring.
We even should/could make `numpy` keg only, with the same reason as for the `cython` formulae.
Users should use `pip` to install numpy.

I am worried though that some users might have been using brewed `numpy` as-is. From this PR on, the behaviour will change and they will have to manually install brewed Python 3.9 or Python 3.10 if they want to use our brewed `numpy`. How do we feel about that change?

MikeMcQuaid commented:
> I reduced the scope to the bare minimun, as asked: Python/Cython 3.10 and 3.9 only.

Thanks ❤️ 

> Once I have migrated all formulae to `cython-39` and `cython-310`, we can drop the unversioned `cython` formulae completely. Both new formulae are for internal usage only, if someone needs `cython`, they should use `pip`, as stated in the keg only reason.

What about keeping it around and keeping it for the default `python` version? I think the naming is 😭 enough that it'd be nice to avoid it where possible.



> Note that for `numpy`, there is no Python runtime dependency anymore. I consider `numpy` as an internal library, as it is only there to help build some formulae that need `numpy` and reduce vendoring.
> We even should/could make `numpy` keg only, with the same reason as for the `cython` formulae.
> Users should use `pip` to install numpy.

👍🏻 



> I am worried though that some users might have been using brewed `numpy` as-is. From this PR on, the behaviour will change and they will have to manually install brewed Python 3.9 or Python 3.10 if they want to use our brewed `numpy`. How do we feel about that change?

Make the change but be prepared to revert if we get 2+ complaints.

iMichka commented:
> What about keeping it around and keeping it for the default python version? I think the naming is 😭 enough that it'd be nice to avoid it where possible.

Yes I can do everything/merge my new `cython-xy` formulae into the original `cython`; which was my initial strategy, but I'll have to add one more audit/style exceptions to brew first. It just felt weird to ship `cython` with a binary based on Python 3.10, and the library part for Python 3.9 and 3.10. Based on @carlocab 's proposal having separate `cython` formulae felt way cleaner, and fits with our style checks. I am wondering if we could not find better names for both formulae.

iMichka commented:
I changed the strategy to provider a better solution.

I added a new `libcython` formula, which does not ship the binary, only the library part. I then migrated all the formulae that could be built with that.

Remains `nest` and `vapoursynth ` which are asking for the binary (though that seems weird, not sure what the build process is doing with that), and cassandra@2.2 and cassandra@3 which are deprecated anyway.

I made sure both `cython` and `libcython` versions are kept in sync.

MikeMcQuaid commented:
> I added a new `libcython` formula, which does not ship the binary, only the library part. I then migrated all the formulae that could be built with that.

Ooo, even better idea.

bayandin commented on Formula/libcython.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def install
+    pythons.each do |python|
+      ENV.prepend_create_path "PYTHONPATH", libexec/Language::Python.site_packages(python)
```
For the second (and subsequent) python `PYTHONPATH`  includes site packages from previous pythons. Should we remove it?

iMichka replied:
I think I got weird errors mixing Python versions, and anyway I want to specifically install `cython` libraries twice in this case, once for each Python version. So resetting the `PYTHONPATH` each time is the thing to do.

bayandin replied:
Ah sorry, probably I worded the question badly (or I didn't get the answer).
I wanted to point out that after building cython for python@3.10 we still have `lib/python3.10/site-packages` in PYTHONPATH. And then we build it for python@3.9 — we're adding `lib/python3.9/site-packages` on top of `lib/python3.10/site-packages` in PYTHONPATH, aren't we?

iMichka replied:
Ha yes right got it. `ENV.prepend_create_path` twice is wrong, you are right. I'll update the PR/test this more later today, this needs to be changed.

iMichka replied:
I checked and this is what we have for the first and second loop:
```
/usr/local/Cellar/libcython/0.29.28/libexec/lib/python3.10/site-packages
/usr/local/Cellar/libcython/0.29.28/libexec/lib/python3.9/site-packages:/usr/local/Cellar/libcython/0.29.28/libexec/lib/python3.10/site-packages
```

As we prepend and the first element in the path is used, all is good. Ideally we could add one step to delete the `PYTHONPATH` at the end of the loop but it does not feel necessary.

So I think the code is doing the right thing.

cho-m replied:
Does `PYTHONPATH` do anything in this case? It is usually used to find other packages, but I don't think it matters when the path is for the package that is currently being installed.

`cython` needs it to create env scripts so executables can find unlinked libraries.
